### PR TITLE
Configuration for snapshot-controller

### LINKED ISSUE #362: [2] Add configuration for snapshot controller
### Proposed feature

Add configuration for enabling the snapshot controller. `csi_snapshot_controller_enabled: true`

Also add configuration for adding volumeSnapshotClasses for each CSI driver.

In rook chart there is templating already for this.

### Proposed alternatives

_No response_

### Additional context

_No response_

### Definition of done

- [x] Snapshot controller can be enabled on all infra providers
- [x] It is possible to create VolumeSnapshotClass for every CSI Driver

### PR DESCRIPTION
**Pull request by robinelastisys**
> [!warning]
> **This is a public repository, ensure not to disclose:**
>
> - [x] personal data beyond what is necessary for interacting with this pull request, nor
> - [x] business confidential information, such as customer names.

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [ ] kind/feature       
- [x] kind/improvement   
- [ ] kind/deprecation   
- [ ] kind/documentation 
- [ ] kind/clean-up      
- [ ] kind/bug           
- [ ] kind/other         

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [ ] kind/admin-change 
- [ ] kind/dev-change   
- [ ] kind/security     
- [ ] [kind/adr]()      

### What does this PR do / why do we need this PR?

- Fixes #362 

#### Information to reviewers

My main question to my reviewers is... Is this enough? The snapshot-controller pod has been enabled for cinder-csi for a very long and it seems like a VolumeSnapshotClass already is defined by default? A clearer definition of done may be needed. Because of this, I would like to have some input on what we want to achieve at the end of the day. :thinking: 

#### Checklist

- [x] Proper commit message prefix on all commits
  
- Change checks:
  - [x] The change is transparent
  - [ ] The change is disruptive
  - [x] The change requires no migration steps
  - [ ] The change requires migration steps
- Metrics checks:
  - [x] The metrics are still exposed and present in Grafana after the change
  - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [x] The logs do not show any errors after the change
- Pod Security Policy checks:
  - [ ] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [x] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Network Policy checks:
  - [x] Any changed pod is covered by Network Policies
  - [ ] The change does not cause any dropped packets in the `NetworkPolicy Dashboard`
- Audit checks:
  - [x] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy
- Falco checks:
  - [x] The change does not cause any alerts to be generated by Falco
- Bug checks:
  - [x] The bug fix is covered by regression tests

### CODE REVIEW

**Comment by robinAwallace** [comment]
It feels like the csi-controller in kubespray is not as feature rich as the upstream chart :thinking: Im unsure if we need everything or if this is enough. 

Also, `volumeSnapshotClasses` is still missing. But that is something you need add to upstream kubespray.

**Comment by robinelastisys** [comment]
A few realizations have been made:
This wont be enough as Upcloud needs a solo config. We wont have to change something in our Openstack settings as that is already fixed by `cinder_csi_enabled: true` - what have been done so far could essentially be removed? What need to be done is to add the configs Upcloud need.

File: kubespray
Comment (Xartos) [inline kubespray:1]: Why was this changed?

**Review by Xartos — COMMENTED** [review]
I assume that we need to have a migration script for existing clusters to get this setup?
To enable the `csi_snapshot_controller_enabled` variable?

**Comment by Elias-elastisys** [comment]
> This wont be enough as Upcloud needs a solo config. We wont have to change something in our Openstack settings as that is already fixed by `cinder_csi_enabled: true` - what have been done so far could essentially be removed? What needs to be done is to add the configs Upcloud need.

I don't quite understand why you say this wont be enough for upcloud. Once [this PR](https://github.com/kubernetes-sigs/kubespray/pull/11303) is merged, all that is needed is a VolumeSnapshotClass and `csi_snapshot_controller_enabled: true` as I've tested and confirmed. Is there any point in using the upcloud specific settings for this? Is it not more clear to use this general snapshot setting to ensure it is always enabled on all providers?

**Comment by robinelastisys** [comment]
> > This wont be enough as Upcloud needs a solo config. We wont have to change something in our Openstack settings as that is already fixed by `cinder_csi_enabled: true` - what have been done so far could essentially be removed? What needs to be done is to add the configs Upcloud need.
> 
> I don't quite understand why you say this wont be enough for upcloud. Once [this PR](https://github.com/kubernetes-sigs/kubespray/pull/11303) is merged, all that is needed is a VolumeSnapshotClass and `csi_snapshot_controller_enabled: true` as I've tested and confirmed. Is there any point in using the upcloud specific settings for this? Is it not more clear to use this general snapshot setting to ensure it is always enabled on all providers?

Your totally right! :+1:

**Comment by robinelastisys** [comment]
This 3 branches go together:

https://github.com/kubernetes-sigs/kubespray/commit/cb9d5906bb9bd0b075dbb5651ae9d6cb7cb26b7e
https://github.com/elastisys/compliantkubernetes-kubespray/pull/367/files
https://github.com/elastisys/kubespray/pull/16

File: config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+etcd_deployment_type: kubeadm
+
+## Repo for UpClouds csi-driver: https://github.com/UpCloudLtd/upcloud-csi
+## To use UpClouds CSI plugin to provision volumes set this value to true
+## Remember to set UPCLOUD_USERNAME and UPCLOUD_PASSWORD
```
Comment (simonklb) [inline config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml:5]: So the CSI driver uses the same credentials as we use to authenticate with when provisioning the infrastructure using Terraform?

File: config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+upcloud_csi_enabled: true
+upcloud_csi_enable_volume_snapshot: true
+
+#upcloud_csi_controller_replicas: 1
+## Override used image tags
+# upcloud_csi_provisioner_image_tag: "v3.1.0"
+# upcloud_csi_attacher_image_tag: "v3.4.0"
+# upcloud_csi_resizer_image_tag: "v1.4.0"
+# upcloud_csi_plugin_image_tag: "alpha"
+# upcloud_csi_node_image_tag: "v2.5.0"
```
Comment (simonklb) [inline config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml:15]: Do you expect us to ever use these? If not let's just clean it up before merging.

File: config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+upcloud_tolerations:
+  - key: elastisys.io/node-type
+    operator: Exists
+    effect: NoSchedule
```
Comment (simonklb) [inline config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml:19]: We don't want Elastisys specifics in this repository.

File: config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+## Storage class options
+expand_persistent_volumes: true
+persistent_volumes_enabled: true
+storage_classes:
+  - name: standard
+    is_default: true
+
+supplementary_addresses_in_ssl_keys: [<loadbalancer-domain>, "kube.ops.CHANGE-ME-BASE-DOMAIN"]
+
+kubeconfig_localhost_ansible_host: true
```
Comment (simonklb) [inline config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml:30]: These are commented out in the OpenStack config: https://github.com/elastisys/compliantkubernetes-kubespray/blob/dcd2c00a724c33f9279f37c97f0762bd6a06314f/config/openstack/group_vars/k8s_cluster/ck8s-k8s-cluster-openstack.yaml#L42-L44

Why is it different here?

File: kubespray
Comment (simonklb) [reply kubespray:1]: I'm also curious about this change, 360 files changed?

File: rook/helmfile.d/values/cluster.yaml.gotmpl [hunk 5]
```diff
@@ -105,3 +105,7 @@ cephBlockPools:
 cephFileSystems: []
 
 cephObjectStores: []
+
+cephBlockPoolsVolumeSnapshotClass:
+  enabled: true
+  isDefault: true
```
Comment (simonklb) [inline rook/helmfile.d/values/cluster.yaml.gotmpl:111]: Does UpCloud not have block storage or why do we use Rook/Ceph here?

**Review by simonklb — CHANGES_REQUESTED** [review]

File: config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml (hunk 2 above)
Comment (Pavan-Gunda) [reply config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml:15]: I'll clean them as I do not think we'll use these unless there's a bug in this version but then that should be caught during QA and fixed by upgrade CSI version.

File: config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml (hunk 1 above)
Comment (Pavan-Gunda) [reply config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml:5]: We use UpCloud main account credentials to run kubespray which is also used by CSI driver.  I'll make it clearer in the docss

File: rook/helmfile.d/values/cluster.yaml.gotmpl (hunk 5 above)
Comment (Pavan-Gunda) [reply rook/helmfile.d/values/cluster.yaml.gotmpl:111]: https://github.com/elastisys/compliantkubernetes-kubespray/issues/362#:~:text=Also%20add%20configuration%20for%20adding%20volumeSnapshotClasses%20for%20each%20CSI%20driver.

The issue description asks that we have config for enabling volumesnapshotclass for all CSI drivers.

File: kubespray
Comment (Xartos) [inline kubespray:1]: There's a lot of changes here, do we really need all these changes?

File: config/common/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml [hunk 6]
```diff
@@ -61,6 +61,8 @@ coredns_additional_error_config: |
 
 kube_profiling: false
 
+csi_snapshot_controller_enabled: true
```
Comment (Xartos) [inline config/common/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml:64]: Will this also create the VolumeSnapshotClass? Or only the controller? Same with upcloud

Comment (Pavan-Gunda) [reply config/common/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml:64]: This will only create the controller. The volumesnapshotclass is created when cinder_csi is enabled, which seems a little weird. I would want volumesnapshotclass created when its controller is enabled. See [here](https://github.com/elastisys/kubespray/blob/81446462cf0bb04411ee7de1c734c8b7559336e0/roles/kubernetes-apps/snapshots/meta/main.yml) and [here](https://github.com/elastisys/kubespray/blob/81446462cf0bb04411ee7de1c734c8b7559336e0/roles/kubernetes-apps/snapshots/cinder-csi/templates/cinder-csi-snapshot-class.yml.j2)

Do we want to fix that in kubespray ? 

In UpCloud's case, a volumesnapshotclass is created when snapshotcontroller is enabled.

File: kubespray
Comment (Pavan-Gunda) [reply kubespray:1]: No, we do not. This kubespray submodule is  synced with the latest upstream master and then snapshot specific commits are on top of it. I'll change that and just add snapshot specific commit.

File: config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml (hunk 3 above)
Comment (Pavan-Gunda) [reply config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml:12]: I'll remove them from here and add it in our templates.

File: config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml (hunk 4 above)
Comment (Pavan-Gunda) [reply config/upcloud/group_vars/k8s_cluster/ck8s-k8s-cluster-upcloud.yaml:18]: I did a quick check in our internal templates and in it, both openstack and upcloud have this config uncommented. I'll uncomment the openstack condig here.

File: config/common/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml (hunk 6 above)
Comment (Xartos) [reply config/common/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml:64]: Alright, so the volumesnapshotclass should already be installed in all our cluster already then? I'm fine with leaving it tbh.

File: kubespray
Comment (Xartos) [reply kubespray:1]: Still seeing a lot of commits, but maybe you didn't finish this yet?

File: config/common/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml (hunk 6 above)
Comment (Pavan-Gunda) [reply config/common/group_vars/k8s_cluster/ck8s-k8s-cluster.yaml:64]: Yes, they're already installed in our clusters. I'll resolve this then.

**Review by viktor-f — COMMENTED** [review]
I see you have a conflict, but otherwise it looks good to me

File: kubespray
Comment (Xartos) [inline kubespray:1]: This still looks a bit off

```console
$ curl https://github.com/elastisys/kubespray/commit/ac2fc294bcc73df01e3ad07836ad5054c06c0e0f
Not Found
```

Have you pushed that commit?

**Comment by Pavan-Gunda** [comment]
I created a new PR with the same changes as this one as I didn't manage to fix the conflicts in the submodule. Please take a look 

https://github.com/elastisys/compliantkubernetes-kubespray/pull/382

**Comment by Pavan-Gunda** [comment]
Closing in favor of https://github.com/elastisys/compliantkubernetes-kubespray/pull/382
